### PR TITLE
build(devtools-view): Node16 resolution with ESM support

### PR DESCRIPTION
**Pull request by DLehenbauer**
[AB#7223](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/7223)

Add ESM build and always build with module Node16.

- .js extensions from ts2esm
- Change no-internal-modules to use standard lint config, then disable specific other places as needed.
- Regorganize test fully under src/test and extract mock to src/test/utils
- Setup jest to run against pre-built files to workaround ts-jest limitation with .cts files
    Use Node16 resolution and build jest tests for CJS
    ts-jest settings are commented out as a precaution.
- Remove dead malformed interfaces (CommonInterfaces.ts)
- Use a less standard build pattern with broad `fluid-tsc` to enable full module<->cjs switch support
    Enable ability to change root package.json "type" with low effort as future investigations may be performed.
    Besides changing that type it is recommended to update projects listed in lint config to reflect what is linted. test/screenshot will be inaccurately linted with "type": "common" at root but not a terrible hit.

### CODE REVIEW

File: packages/tools/devtools/devtools-view/package.json
Comment (tylerbutler) [inline packages/tools/devtools/devtools-view/package.json:166]: Should be able to remove this since ^tsc is a sub dependency (not sure what else to call it) through the utils.cjs task.

File: packages/tools/devtools/devtools-view/package.json
Comment (tylerbutler) [inline packages/tools/devtools/devtools-view/package.json:171]: Same here.

File: packages/tools/devtools/devtools-view/package.json
Comment (tylerbutler) [inline packages/tools/devtools/devtools-view/package.json:184]: Does this also need the dependent ESM output? I.e. ^build:esnext?

File: packages/tools/devtools/devtools-view/package.json
Comment (tylerbutler) [inline packages/tools/devtools/devtools-view/package.json:181]: Does this need the local source to be built too? That is, `tsc`?

**Comment by tylerbutler** [comment]
Great changes. This was a gnarly one!

File: packages/tools/devtools/devtools-view/tsconfig.json
Comment (alexvy86) [inline packages/tools/devtools/devtools-view/tsconfig.json:7]: Probably worth a comment here (maybe inside the `outDir` string) to say that this is just a base file for the esm/cjs tsconfig files?

File: packages/tools/devtools/devtools-view/src/test/jest/tsconfig.esm.json
Comment (alexvy86) [inline packages/tools/devtools/devtools-view/src/test/jest/tsconfig.esm.json:8]: Shouldn't this and its sibling file have "jest" types here, like the previous tsconfig?

File: packages/tools/devtools/devtools-view/src/components/SplitPane.cts
Comment (alexvy86) [inline packages/tools/devtools/devtools-view/src/components/SplitPane.cts:1]: Can you clarify why we need this file? I see the import in TelemetryView was updated to grab `SplitPane` from here instead of directly from `react-split-pane` but don't understand why it was necessary.

**Review by alexvy86 — COMMENTED** [review]
Overall it makes sense to me, but I'd like to wait for @Josmithr to review as well.

File: packages/tools/devtools/devtools-view/src/test/jest/NoDevtoolsErrorBar.test.tsx
Comment (jason-ha) [reply packages/tools/devtools/devtools-view/src/test/jest/NoDevtoolsErrorBar.test.tsx:11]: Yes. There is a lint rule for this. It doesn't seem super consistently visible, though. From the original pattern + full ESM build we moved this import to a .cts file. Then the lint rule showed up. Have a look at [ce50831](https://github.com/microsoft/FluidFramework/pull/19683/commits/ce50831b7ecaf28c419d21b77e0bd09b5219b67a) where I was clued in that named import use was the way to go.

File: packages/tools/devtools/devtools-view/src/test/jest/tsconfig.esm.json
Comment (jason-ha) [reply packages/tools/devtools/devtools-view/src/test/jest/tsconfig.esm.json:8]: It should. jest typing appears to be transitively picked up with import of @testing-library/jest-dom, but we should be explicit. I'll patch.

File: packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx
Comment (Josmithr) [inline packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx:24]: Any particular reason we moved these from the config? We made the explicit decision to allow imports from this specific path, and have taken a `~` dependency on the library to try to ensure we don't get hit with breaking changes here.

Unless there is a specific reason we want to change the pattern, I think I prefer how it was for a couple few reasons:
1. Centralized location for documentation
2. Less eslint-disable cruft
3. For devs making changes / adding new functionality, it prevents them having to be expressly aware that we've made this allowance - they can just import without issue. I'd like to prevent devs seeing the linter error and assuming they can't use `react-components/unstable`.

File: packages/tools/devtools/devtools-view/src/components/SplitPane.cts
Comment (Josmithr) [reply packages/tools/devtools/devtools-view/src/components/SplitPane.cts:1]: And if it is needed, let's capture that in docs in this file please.

File: packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx
Comment (jason-ha) [reply packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx:24]: Yes. Commented in the description:
> Change no-internal-modules to use standard lint config, then disable specific other places as needed.
Maybe there is a better pattern to combine rules for `no-internal-modules`. But I wasn't enamored with a general allowance for something called unstable - seemed like that should allowed on a case-by-case basis. I will defer to your preference.

Comment (Josmithr) [reply packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx:24]: I take it the base config already contains exceptions, and this package's config was effectively removing them? I guess I'd still prefer that, and add whatever other exceptions we need to the override.

In this particular case, we are opting into "experimental" features from FluentUI, and that was a conscious decision. I believe most of the experimental features we opted into have been promoted to stable in more recent releases, but I suspect upgrading would be non-trivial and should be tracked separately. At that point we might be able to remove the exception altogether. I'll add something to Team Rohan's backlog for that.

File: packages/tools/devtools/devtools-view/src/components/SplitPane.cts
Comment (Josmithr) [inline packages/tools/devtools/devtools-view/src/components/SplitPane.cts:6]: Thanks!!

File: packages/tools/devtools/devtools-view/tsconfig.json
Comment (Josmithr) [inline packages/tools/devtools/devtools-view/tsconfig.json:8]: Nit
```suggestion
		// emitting; so, set an invalid outDir.
```

**Review by Josmithr — APPROVED** [review]

File: packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx
Comment (jason-ha) [reply packages/tools/devtools/devtools-view/src/components/ContainerSummaryView.tsx]: Correct. Restored with rule cloning.

**Pull request merged**
